reference/best-practices/java-app.md

```diff
@@ -0,0 +1,362 @@
+```
+
+### `ExecutorType`
+
+You can choose `ExecutorType` during `openSession`. MyBatis supports three types of executors:
+
+- Simple: The Prepared Statements are called to JDBC for each execution (if the JDBC configuration item `cachePrepStmts` is enabled, repeated Prepared Statements will be reused)
+- Reuse: The Prepared Statements are cached in `executor`, so that you can reduce duplicate calls for Prepared Statements without using the JDBC `cachePrepStmts`
+- Batch: Each update statement will only be executed with `ExecuteBatch()` called when `addBatch()` is used for a query or a commit. If `rewriteBatchStatements` is enabled in the JDBC layer, it will try to rewrite the statements. If not, the statements will be sent one by one.
+
+Usually, the default value of `ExecutorType` is `Simple`. You need to change `ExecutorType` when calling `openSession`. If it is batch execution, you might find that in a transaction the update or insert statements are executed pretty fast, but it is slower when reading data or committing the transaction. This is actually normal, so you need to note this when troubleshooting slow SQL queries.
+
+## Spring Transaction
+
+In the real world, applications might use [Spring Transaction](https://docs.spring.io/spring/docs/4.2.x/spring-framework-reference/html/transaction.html) and AOP aspects to start and stop transactions.
+
+By adding the `@Transactional` annotation to the method definition, AOP starts the transaction before the method is called, and commit the transaction before the method returns the result. If your application has a similar need, you can find `@Transactional` in code to determine when the transaction is started and closed.
+
+Pay attention to a special case of embedding. If it occurs, Spring will behave differently based on the [Propagation](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/annotation/Propagation.html) configuration. Because TiDB does not support savepoint, nested transactions are not supported yet.
+
+## Misc
+
+### Troubleshooting tools
+
+Using the powerful troubleshooting tools of JVM is recommended when an issue occurs in your Java application and you do not know the business logic. Here are a few common tools:
+
+#### jstack
+
+[jstack](https://docs.oracle.com/javase/7/docs/technotes/tools/share/jstack.html) is similar to pprof/goroutine in Go, which can easily troubleshoot the process stuck issue.
+
+By executing `jstack pid`, you can output the IDs and stack information of all threads in the target process. There is only the Java stack in the default output. If you want to output the C++ stack in the JVM at the same time, add the `-m` option.
+
+By using jstack multiple times, you can easily locate the stuck issue (such as updating only through Mybatis BatchExecutor flush) or the deadlock issue (such as the failure to send an SQL statement because test programs are preempting a lock in the application)
```

Comment:
how about

```
By using jstack multiple times, you can easily locate the stuck issue (such slow query from application's view due to using Batch ExecutorType in Mybatis) or the application deadlock issue (such as application do not send any SQL statements because it is preempting a lock before sending)
```

Reply:
Updated.